poly-graph-persistent/src/Data/Graph/HGraph/Persistent/TH.hs

```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Data.Graph.HGraph.Persistent.TH
+  ( UniquenessCheck(..)
+  , mkUniquenessChecks
+  , mkUniquenessCheck
+  ) where
+
+import Control.Arrow ((&&&))
+import Data.Char (toLower, toUpper)
+import Data.List.NonEmpty (nonEmpty)
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Maybe (mapMaybe)
+import Data.Monoid ((<>))
+import Data.Text (Text, unpack, cons, uncons)
+import Database.Persist
+import Database.Persist.Quasi (nullable)
+import Database.Persist.TH
+import Language.Haskell.TH
+
+class UniquenessCheck a where
+  couldCauseUniquenessViolation :: a -> a -> Bool
+
+mkUniquenessChecks :: MkPersistSettings -> [EntityDef] -> Q [Dec]
```

Comment:
I'll add a comment. To add a manual instance, I'd have to separate one entity out from the `persistLowerCase` quasi-quoter block since this is designed to be used with `share`.